File: app/src/main/java/blue/koenig/kingsfinances/model/database/StandingOrderTable.java
package blue.koenig.kingsfinances.model.database;

import android.content.ContentValues;
import android.database.Cursor;
import android.database.sqlite.SQLiteDatabase;

import com.koenig.commonModel.Frequency;
import com.koenig.commonModel.finance.BookkeepingEntry;
import com.koenig.commonModel.finance.Expenses;
import com.koenig.commonModel.finance.StandingOrder;

import org.joda.time.DateTime;

import java.util.Arrays;
import java.util.Collection;
import java.util.List;

/**
 * Created by Thomas on 30.11.2017.
 */

public class StandingOrderTable extends BookkeepingTable<StandingOrder> {
    public static final String NAME = "standing_order_table";
    private static final String FIRST_DATE = "first_date";
    private static final String END_DATE = "end_date";
    private static final String FREQUENCY = "frequency";
    private static final String FREQUENCY_FACTOR = "frequency_factor";
    private static final String EXECUTED_EXPENSES = "executed_expenses";

    public StandingOrderTable(SQLiteDatabase database) {
        super(database);
    }

    @Override
    protected String getBookkeepingTableSpecificCreateStatement() {
        return "," + FIRST_DATE + " LONG, "
                + END_DATE + " LONG, "
                + FREQUENCY + " TEXT,"
                + FREQUENCY_FACTOR + " INT, "
                + EXECUTED_EXPENSES + " TEXT ";
    }

    @Override
    protected void setBookkeepingItem(ContentValues values, StandingOrder item) {
        values.put(FIRST_DATE, dateTimeToValue(item.getFirstDate()));
        values.put(END_DATE, dateTimeToValue(item.getEndDate()));
        values.put(FREQUENCY, item.getFrequency().name());
        values.put(FREQUENCY_FACTOR, item.getFrequencyFactor());
        values.put(EXECUTED_EXPENSES, buildStringList(item.getExecutedExpenses()));
    }

    @Override
    protected StandingOrder getBookkeepingItem(BookkeepingEntry entry, Cursor cursor) {
        DateTime firstDate = getDateTime(cursor, FIRST_DATE);
        DateTime endDate = getDateTime(cursor, END_DATE);
        Frequency frequency = getEnum(cursor, FREQUENCY, Frequency.class);
        int frequencyFactor = getInt(cursor, FREQUENCY_FACTOR);
        List<String> executedExpenses = getStringList(cursor, EXECUTED_EXPENSES);
        return new StandingOrder(entry, firstDate, endDate, frequency, frequencyFactor, executedExpenses);
    }




    @Override
    protected Collection<? extends String> getBookkeepingColumnNames() {
        return Arrays.asList(FIRST_DATE, END_DATE, FREQUENCY, FREQUENCY_FACTOR, EXECUTED_EXPENSES);
    }

    @Override
    public String getTableName() {
        return NAME;
    }
}

File: app/src/main/java/blue/koenig/kingsfinances/model/database/StandingOrderTable.kt
package blue.koenig.kingsfinances.model.database

import android.content.ContentValues
import android.database.Cursor
import android.database.sqlite.SQLiteDatabase
import android.database.sqlite.SQLiteStatement
import com.koenig.commonModel.Byteable
import com.koenig.commonModel.Frequency
import com.koenig.commonModel.finance.BookkeepingEntry
import com.koenig.commonModel.finance.StandingOrder
import com.koenig.commonModel.toInt
import org.joda.time.LocalDate
import java.sql.SQLException
import java.util.*
import java.util.concurrent.locks.ReentrantLock

/**
 * Created by Thomas on 30.11.2017.
 */

class StandingOrderTable(database: SQLiteDatabase, lock: ReentrantLock) : BookkeepingTable<StandingOrder>(database, lock) {

    override val tableName: String
        get() = NAME

    override val bookkeepingTableSpecificCreateStatement: String = ",$FIRST_DAY INT, $END_DAY INT, $FREQUENCY TEXT,$FREQUENCY_FACTOR INT, $EXECUTED_EXPENSES BLOB "

    override fun setBookkeepingItem(values: ContentValues, item: StandingOrder) {
        values.put(FIRST_DAY, item.firstDate.toInt())
        values.put(END_DAY, item.endDate.toInt())
        values.put(FREQUENCY, item.frequency.name)
        values.put(FREQUENCY_FACTOR, item.frequencyFactor)
        values.put(EXECUTED_EXPENSES, Byteable.getBytesShortMap(item.executedExpenses))
    }

    override fun bindBookkeepingItem(statement: SQLiteStatement, map: Map<String, Int>, item: StandingOrder) {
        statement.bindLong(map[FIRST_DAY]!!, item.firstDate.toInt().toLong())
        statement.bindLong(map[END_DAY]!!, item.endDate.toInt().toLong())
        statement.bindString(map[FREQUENCY]!!, item.frequency.name)
        statement.bindLong(map[FREQUENCY_FACTOR]!!, item.frequencyFactor.toLong())
        statement.bindBlob(map[EXECUTED_EXPENSES]!!, Byteable.getBytesShortMap(item.executedExpenses))
    }

    override fun getBookkeepingItem(entry: BookkeepingEntry, cursor: Cursor): StandingOrder {
        val firstDate = getLocalDate(cursor, FIRST_DAY)
        val endDate = getLocalDate(cursor, END_DAY)
        val frequency = getEnum(cursor, FREQUENCY, Frequency::class.java)
        val frequencyFactor = getInt(cursor, FREQUENCY_FACTOR)
        val executedExpenses = getExecutedExpenses(cursor, EXECUTED_EXPENSES)
        return StandingOrder(entry, firstDate, endDate, frequency, frequencyFactor, executedExpenses)
    }

    fun getExecutedExpenses(cursor: Cursor, expenses: String): MutableMap<LocalDate, String> {
        return Byteable.byteToShortMap(cursor.getBlob(cursor.getColumnIndex(expenses)))
    }


    override val bookkeepingColumnNames: Collection<String> = Arrays.asList(FIRST_DAY, END_DAY, FREQUENCY, FREQUENCY_FACTOR, EXECUTED_EXPENSES)

    @Throws(SQLException::class)
    fun addExpensesToStandingOrders(standingOrderId: String, expensesId: String, day: LocalDate) {
        val standingOrder = getFromId(standingOrderId)!!
        standingOrder.executedExpenses.set(day, expensesId)
        update(standingOrderId, mutableListOf(EXECUTED_EXPENSES), { statement, columnsMap -> statement.bindBlob(columnsMap[EXECUTED_EXPENSES]!!, Byteable.getBytesShortMap(standingOrder.executedExpenses)) })
    }

    companion object {
        val NAME = "standing_order_table"
        private val FIRST_DAY = "first_day"
        private val END_DAY = "end_day"
        private val FREQUENCY = "frequency"
        private val FREQUENCY_FACTOR = "frequency_factor"
        private val EXECUTED_EXPENSES = "executed_expenses"
    }
}

